sycl/source/detail/pi.cpp

```diff
@@ -225,19 +225,34 @@ std::string memFlagsToString(pi_mem_flags Flags) {
 std::shared_ptr<plugin> GlobalPlugin;
 
 // Find the plugin at the appropriate location and return the location.
-bool findPlugins(std::vector<std::pair<std::string, backend>> &PluginNames) {
+std::vector<std::pair<std::string, backend>> findPlugins() {
+  std::vector<std::pair<std::string, backend>> PluginNames;
+
   // TODO: Based on final design discussions, change the location where the
   // plugin must be searched; how to identify the plugins etc. Currently the
   // search is done for libpi_opencl.so/pi_opencl.dll file in LD_LIBRARY_PATH
   // env only.
   //
+  const char *OpenCLPluginName =
+      SYCLConfig<SYCL_OVERRIDE_PI_OPENCL>::get()
+          ? SYCLConfig<SYCL_OVERRIDE_PI_OPENCL>::get()
+          : __SYCL_OPENCL_PLUGIN_NAME;
+  const char *L0PluginName =
+      SYCLConfig<SYCL_OVERRIDE_PI_LEVEL_ZERO>::get()
+          ? SYCLConfig<SYCL_OVERRIDE_PI_LEVEL_ZERO>::get()
+          : __SYCL_LEVEL_ZERO_PLUGIN_NAME;
+  const char *CUDAPluginName = SYCLConfig<SYCL_OVERRIDE_PI_CUDA>::get()
+                                   ? SYCLConfig<SYCL_OVERRIDE_PI_CUDA>::get()
+                                   : __SYCL_CUDA_PLUGIN_NAME;
+  const char *ROCMPluginName = SYCLConfig<SYCL_OVERRIDE_PI_ROCM>::get()
+                                   ? SYCLConfig<SYCL_OVERRIDE_PI_ROCM>::get()
+                                   : __SYCL_ROCM_PLUGIN_NAME;
   device_filter_list *FilterList = SYCLConfig<SYCL_DEVICE_FILTER>::get();
   if (!FilterList) {
-    PluginNames.emplace_back(__SYCL_OPENCL_PLUGIN_NAME, backend::opencl);
-    PluginNames.emplace_back(__SYCL_LEVEL_ZERO_PLUGIN_NAME,
-                             backend::level_zero);
-    PluginNames.emplace_back(__SYCL_CUDA_PLUGIN_NAME, backend::cuda);
-    PluginNames.emplace_back(__SYCL_ROCM_PLUGIN_NAME, backend::rocm);
+    PluginNames.emplace_back(OpenCLPluginName, backend::opencl);
+    PluginNames.emplace_back(L0PluginName, backend::level_zero);
+    PluginNames.emplace_back(CUDAPluginName, backend::cuda);
+    PluginNames.emplace_back(ROCMPluginName, backend::rocm);
```

Comment:
Wouldn't it be better to allow the user to specify the custom library PATH instead of just custom library name?

Reply:
That's exactly what I'm doing. I have a very simple C++ application, that tries to locate itself with:
```c++
std::array<char, 1024> buf;
readlink("/proc/self/exe", buf.data(), buf.size());
mExecutablePath = std::filesystem::path{buf.data()};
```

And then use this information to locate my custom plugin and set the env variables.